lib/class-wp-theme-json-schema-gutenberg.php

```diff
@@ -0,0 +1,195 @@
+	 * This will be applied at both the defaults and individual block levels.
+	 */
+	const V1_TO_V2_RENAMED_PATHS = array(
+		'border.customRadius'         => 'border.radius',
+		'spacing.customMargin'        => 'spacing.margin',
+		'spacing.customPadding'       => 'spacing.padding',
+		'typography.customLineHeight' => 'typography.lineHeight',
+	);
+
+	/**
+	 * Function that migrates a given theme.json structure to the last version.
+	 *
+	 * @since 5.9.0
+	 *
+	 * @param array $theme_json The structure to migrate.
+	 *
+	 * @return array The structure in the last version.
+	 */
+	public static function migrate( $theme_json ) {
+		if ( ! isset( $theme_json['version'] ) ) {
+			$theme_json = array(
+				'version' => WP_Theme_JSON::LATEST_SCHEMA,
+			);
+		}
+
+		if ( 1 === $theme_json['version'] ) {
+			$theme_json = self::migrate_v1_to_v2( $theme_json );
+		}
+
+		if ( 2 === $theme_json['version'] ) {
+			$theme_json = self::migrate_deprecated_lightbox_behaviors( $theme_json );
+		}
+
+		return $theme_json;
+	}
+
+	/**
+	 * Removes the custom prefixes for a few properties
+	 * that were part of v1:
+	 *
+	 * 'border.customRadius'         => 'border.radius',
+	 * 'spacing.customMargin'        => 'spacing.margin',
+	 * 'spacing.customPadding'       => 'spacing.padding',
+	 * 'typography.customLineHeight' => 'typography.lineHeight',
+	 *
+	 * @since 5.9.0
+	 *
+	 * @param array $old Data to migrate.
+	 *
+	 * @return array Data without the custom prefixes.
+	 */
+	private static function migrate_v1_to_v2( $old ) {
+		// Copy everything.
+		$new = $old;
+
+		// Overwrite the things that changed.
+		if ( isset( $old['settings'] ) ) {
+			$new['settings'] = self::rename_paths( $old['settings'], self::V1_TO_V2_RENAMED_PATHS );
+		}
+
+		// Set the new version.
+		$new['version'] = 2;
+
+		return $new;
+	}
+
+
+	/**
+	 * Migrate away from the previous syntax that used a top-level "behaviors" key
+	 * in the `theme.json` to a new "lightbox" setting.
+	 *
+	 * @since 16.7.0
+	 *
+	 * @param array $old Data with (potentially) behaviors.
+	 * @return array Data with behaviors removed.
+	 */
+	private static function migrate_deprecated_lightbox_behaviors( $old ) {
+		// Copy everything.
+		$new = $old;
+
+		// Migrate the old behaviors syntax to the new "lightbox" syntax.
+		if ( isset( $old['behaviors']['blocks']['core/image']['lightbox']['enabled'] ) ) {
+			_wp_array_set(
+				$new,
+				array( 'settings', 'blocks', 'core/image', 'lightbox', 'enabled' ),
+				$old['behaviors']['blocks']['core/image']['lightbox']['enabled']
+			);
+		}
+
+		// Migrate the behaviors setting to the new syntax. This setting controls
+		// whether the Lightbox UI shows up in the block editor.
+		if ( isset( $old['settings']['blocks']['core/image']['behaviors']['lightbox'] ) ) {
+			_wp_array_set(
+				$new,
+				array( 'settings', 'blocks', 'core/image', 'lightbox', 'allowEditing' ),
+				$old['settings']['blocks']['core/image']['behaviors']['lightbox']
+			);
+		}
```

Comment:
This is looking good to me, but I think we can add a note saying that this method doesn't need to be ported to core and can be removed in a later Gutenberg version.

I had forgotten that we actually did have an instance of a migration that never made it into core. You can see it in #36155 and #36182. I was also going to suggest showing some sort of warning, but it looks like that wasn't done in the past. Probably just a dev note for the release.